components/experiment-creation/ExperimentForm.tsx

```diff
@@ -97,179 +110,204 @@ const ExperimentForm = ({
-                    Begin
-                  </Button>
-                </div>
-              </div>
-              <div className={classes.formPart} ref={formPartBasicInfoRef} style={{ width: constrictorSizes.width }}>
-                <Paper className={classes.paper}>
-                  <BasicInfo />
-                </Paper>
-                <div className={classes.formPartActions}>
-                  <Button onClick={prevStage}>Previous</Button>
-                  <Button onClick={nextStage} variant='contained' color='primary'>
-                    Next
-                  </Button>
-                </div>
-              </div>
-              <div className={classes.formPart} ref={formPartAudienceRef} style={{ width: constrictorSizes.width }}>
-                <Paper className={classes.paper}>
-                  <Audience formikProps={formikProps} indexedSegments={indexedSegments} />
-                </Paper>
-                <div className={classes.formPartActions}>
-                  <Button onClick={prevStage}>Previous</Button>
-                  <Button onClick={nextStage} variant='contained' color='primary'>
-                    Next
-                  </Button>
-                </div>
-              </div>
-              <div className={classes.formPart} ref={formPartMetricsRef} style={{ width: constrictorSizes.width }}>
-                <Paper className={classes.paper}>
-                  <Metrics />
-                </Paper>
-                <div className={classes.formPartActions}>
-                  <Button onClick={prevStage}>Previous</Button>
-                  <Button onClick={nextStage} variant='contained' color='primary'>
-                    Next
-                  </Button>
-                </div>
-              </div>
-              <div className={classes.formPart} ref={formPartSubmitRef} style={{ width: constrictorSizes.width }}>
-                <Paper className={classes.paper}>
-                  <Typography variant='h4' gutterBottom>
-                    Confirm and Submit Your Experiment
-                  </Typography>
-                  <Typography variant='body2' gutterBottom>
-                    Now is a good time to{' '}
-                    <Link href='https://github.com/Automattic/abacus/wiki'>
-                      check our wiki&apos;s experiment creation checklist
-                    </Link>{' '}
-                    and confirm everything is in place.
-                  </Typography>
+        const prevStage = () => {
+          if (currentStageIndex === 0) {
+            return
+          }
+
+          const prevStageIndex = currentStageIndex - 1
+          changeStage(stages[prevStageIndex].id)
+        }
+        const nextStage = () => {
+          if (stages.length <= currentStageIndex) {
+            return
+          }
+
+          const nextStageIndex = currentStageIndex + 1
+          changeStage(stages[nextStageIndex].id)
+        }
```

Comment:
Looks like some of these functions can be simplified, e.g., 

```suggestion
        const prevStage = () => {
          if (currentStageIndex > 0) {
            changeStage(stages[currentStageIndex - 1].id)
          }
        }
        const nextStage = () => {
          if (currentStageIndex < stages.length) {
            changeStage(stages[currentStageIndex + 1].id)
          }
        }
```

There's also a bug in the current `nextStage()` because if `currentStageIndex` is `stages.length - 1`, it tries to increment to `stages.length`.

Reply:
Nice catching that bug!

**I am happy to go through with these changes but I do code like this purposefully:**

1. Using short-circuit return instead of a wrapped if.

    There is a subtle but important difference between these two - the way they return:
    - A short-circuit explicitly returns under explicit conditions.
    - The wrapped if implicitly returns under implicit conditions.

    I believe short-circuit returns greatly improve readability - both by making control-flow more linear, by being a pattern that is familiar. I also think it adds an important layer of semantics (We skip this function if X vs we run this code if X). I believe it also improves later code changes - what if things get expanded, what if the if statement gets changed.

    I tend to use this pattern a lot, it is more generally known as the 'return early, return often' pattern, it is not always the best choice but I think it should be the default.
    Plenty of discussion around the web about it.
    And you will see it in my Auth code refactoring soon to be split into a seperate PR.

2. Using intermediate variables

     I do like to seperate out even small operations, I think it helps readability.

Reply:
I think it really depends on the code in question. In my opinion, here both practices reduce readability:

* The short-circuit return is often useful for long functions. But here I find that it adds more mental work, e.g., reading the body of the original `prevStage()` after the `if`, I need to remember that I'm now in the `currentStageIndex !== 0` case, i.e., invert the condition in my mind.
* Single-use intermediate variables are useful when the evaluated expression is complicated. However, each new variable introduces another thing to keep in mind. Taking `prevStage()` as an example again, having to remember what `prevStageIndex` is for one line can be seen as reducing readability.

With the proposed changes, I think that it's much clearer that the functions are called for their side-effects (no `return`s) and that the `currentStageIndex` is checked correctly (which wasn't the case previously).

>  Plenty of discussion around the web about it.

Yeah, plenty of opinions either way. Here a couple from Spartan programming on [vertical complexity](https://github.com/SpartanRefactoring/Main/wiki/Vertical-complexity) and [variable elimination](https://github.com/SpartanRefactoring/Main/wiki/Variable-elimination). I think that the best answer is: It depends. :slightly_smiling_face: